core/src/main/java/de/jplag/merging/MatchMerging.java

```diff
@@ -0,0 +1,134 @@
+package de.jplag.merging;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import de.jplag.JPlagComparison;
+import de.jplag.JPlagResult;
+import de.jplag.Match;
+import de.jplag.Submission;
+import de.jplag.Token;
+import de.jplag.options.JPlagOptions;
+
+/**
+ * This class implements match merging against obfuscation. Based on configurable parameters MergeBuffer and
+ * SeperatingThreshold, it alters prior results and merges all neighboring matches that fit the specified thresholds.
+ * When neighboring matches get merged they become one and the tokens separating them get removed from the submission
+ * clone. MergeBuffer describes how shorter a match can be than the Minimum Token Match. SeperatingThreshold describes
+ * how many tokens can be between two neighboring matches. Both are set in {@link JPlagOptions} as
+ * {@link MergingParameters} and default to 0 (which deactivates merging).
+ */
+public class MatchMerging {
+    private int minimumTokenMatch;
+    private Submission leftSubmission;
+    private Submission rightSubmission;
+    private List<Match> globalMatches;
+    private List<List<Match>> neighbors;
+    private int seperatingThreshold;
+    private JPlagResult result;
+    private List<JPlagComparison> comparisons;
+    private JPlagOptions options;
+
+    /**
+     * Constructor for class MatchMerging
+     * @param result is the initially computed result object
+     * @param options encapsulates the adjustable options
+     */
+    public MatchMerging(JPlagResult result, JPlagOptions options) {
+        this.result = result;
+        this.comparisons = new ArrayList<>(result.getAllComparisons());
+        this.options = options;
+        this.minimumTokenMatch = options.minimumTokenMatch();
+        this.seperatingThreshold = options.mergingParameters().seperatingThreshold();
+    }
+
+    /**
+     * Runs the internal match merging pipeline. It computes neighboring matches, merges them based on
+     * {@link MergingParameters} and removes the merge buffer afterwards.
+     * @return JPlagResult containing the merged matches
+     */
+    public JPlagResult run() {
+        for (int i = 0; i < comparisons.size(); i++) {
+            leftSubmission = comparisons.get(i).firstSubmission().copy();
+            rightSubmission = comparisons.get(i).secondSubmission().copy();
+            globalMatches = new ArrayList<>(comparisons.get(i).matches());
+            globalMatches.addAll(comparisons.get(i).ignoredMatches());
+            computeNeighbors();
+            mergeNeighbors();
+            removeBuffer();
+            comparisons.set(i, new JPlagComparison(leftSubmission, rightSubmission, globalMatches, new ArrayList<>()));
+
+        }
+        return new JPlagResult(comparisons, result.getSubmissions(), result.getDuration(), options);
```

Comment:
Could you explain what you mean or want to be changed?

Reply:
Currently you instantiate the new result, with the duration from the old result. To my understanding (correct me if I'm wrong) this means this is just the time from the GreedyStringTiling.
If the MatchMerging takes a long time especially for larger submissions sets, I think it would be usefull to add this time to the total duration

Reply:
I've never considered the duration. Should not be too hard to increase it in match merging.